### ISSUE TITLE
Bidirectional RNN

### ISSUE THREAD

**Issue by OlaRonning** [deepmind/dm-haiku#298]
I'm working on a project where we needed a bidirectional RNN. I've sketched a solution [here](https://gist.github.com/OlaRonning/2794846b19539e8296c0b5234d883c51), based on PyTorch bi-RNNs. However, I would like to get some help with handling the reversing of padded sequences in a manner appropriate for haiku. PyTorch gets around this with a [data wrapper](https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.pack_padded_sequence.html), which could also make sense in haiku. 

I highlighted the input flipping in the linked gist with a to-do comment. 

If bidirectional RNNs are of interest and we can agree on an appropriate design, I'll write a test sweep and open a PR.

**Comment by OlaRonning** [comment]
I had swapped the time axis and batch axis. The [new implementation](https://gist.github.com/OlaRonning/2794846b19539e8296c0b5234d883c51) resolves this; however, it requires linear sequential steps in the depth of the biRNN. 
The implementation currently only works for Vanilla RNN. I'll resolve this and reopen the issue.
